binder/src/main/java/io/grpc/binder/AsyncSecurityPolicy.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.binder;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.grpc.ExperimentalApi;
+import io.grpc.Status;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.CheckReturnValue;
+
+/**
+ * Decides whether a given Android UID is authorized to access some resource.
+ *
+ * <p>This class provides the asynchronous version of {@link SecurityPolicy}, allowing
+ * implementations of authorization logic that involves slow or asynchronous calls without
+ * necessarily blocking the calling thread.
+ *
+ * @see SecurityPolicy
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/10566")
+@CheckReturnValue
+public abstract class AsyncSecurityPolicy extends SecurityPolicy {
+
+/**
+ * @deprecated Prefer {@link #checkAuthorizationAsync(int)} for async or slow calls or subclass
+ *     {@link SecurityPolicy} directly for quick, synchronous implementations.
+ */
+@Override
+@Deprecated
+public final Status checkAuthorization(int uid) {
+    try {
+        return checkAuthorizationAsync(uid).get();
+    } catch (ExecutionException e) {
+        return Status.fromThrowable(e);
```

Comment:
Thinking this through some more, I wonder if we should throw an UnsupportedOperationException error here instead. 

Otherwise we have no way of handling the possibility of a transient failure (which shouldn't be cached).

Reply:
Hard to say. This code won't really be used for long, only if someone starts using `AsyncSecurityPolicy` right as-is right after this PR is merged. On the next one, I'll make `TransportAuthorizationState.checkAuthorization` call the async method, so I believe there will no longer be any internal caller anymore (though, of course, integrators can always still call it, even though they shouldn't).

Also added a couple more unit tests to handle cancelled vs interrupted futures. I'm re-interrupting the current thread based on http://go/java-practices/interruption#interruptedexception.

Reply:
Sorry to be annoying, but I think it'll be simpler if we just leave this file out of the CL for now, and only add it in the follow up when we also include support for it. Then the question of what we do about this method becomes much much easier.

Reply:
Done. I've also moved the new unit tests out since they required `AsyncSecurityPolicy`.